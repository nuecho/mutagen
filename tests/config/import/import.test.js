const INVALID_TENANT_CONFIGURATION_PATH = getResourcePath("import/switch-with-invalid-tenant-config.json");
const MULTIPLE_OBJECTS_CONFIGURATION_PATH = getResourcePath("import/multiple-objects-config.json");
const REPEATED_KEYS_CONFIGURATION_PATH = getResourcePath("import/same-key-switches-config.json");
const CYCLE_CONFIGURATION_PATH = getResourcePath("import/cycle-config.json");
const MISSING_DEPENDENCY_CONFIGURATION_PATH = getResourcePath("import/missing-dependency-config.json");
const ACCEPT_CHANGES_CONFIGURATION_PATH = getResourcePath("import/accept-changes-config.json");
const REJECT_CHANGES_CONFIGURATION_PATH = getResourcePath("import/reject-changes-config.json");
const SWITCH_DEPENDENCIES_CONFIGURATION_PATH = getResourcePath(
  "config-objects/switch-dependencies-config.json"
);

test(`[mutagen config import --help] should print import usage`, () => {
  const { code, output } = mutagen(`config import --help`);

  expect(output).toMatchSnapshot("import usage");
  expect(code).toBe(0);
});

test(`[mutagen config import] should fail when file to import does not exist`, () => {
  const { code, output } = mutagen(`config import --auto-confirm /tmp/foo`);

  expect(output).toMatchSnapshot("no such file or directory");
  expect(code).not.toBe(0);
});

test(`[mutagen config import file] should not import the object if it has an invalid tenant`, () => {
  mutagen(`config import --auto-confirm ${SWITCH_DEPENDENCIES_CONFIGURATION_PATH}`);
  const { code, output } = mutagen(`config import --auto-confirm ${INVALID_TENANT_CONFIGURATION_PATH}`);

  expect(output).toMatchSnapshot("no object has been imported");
  expect(code).toBe(1);
});

test(`[mutagen config import file] should import multiple objects of different types`, () => {
  const { code, output } = mutagen(`config import --auto-confirm ${MULTIPLE_OBJECTS_CONFIGURATION_PATH}`);

  expect(output).toMatchSnapshot("every object has been imported");
  expect(code).toBe(0);
});

test(`[mutagen config import file] should only import objects with a unique key`, () => {
  const { code, output } = mutagen(`config import --auto-confirm ${REPEATED_KEYS_CONFIGURATION_PATH}`);

  expect(output).toMatchSnapshot("only one object with the key has been imported");
  expect(code).toBe(0);
});

test(`[mutagen config import file] should not import any object if there's a cycle between object dependencies`, () => {
  const { code, output } = mutagen(`config import --auto-confirm ${CYCLE_CONFIGURATION_PATH}`);

  expect(output).toMatchSnapshot("cycle detected");
  expect(code).toBe(1);
});

test(`[mutagen config import file] should not import any object if some dependencies cannot be resolved`, () => {
  const { code, output } = mutagen(`config import --auto-confirm ${MISSING_DEPENDENCY_CONFIGURATION_PATH}`);

  expect(output).toMatchSnapshot("found missing dependency");
  expect(code).toBe(1);
});

test(`[mutagen config import file] should not import the objects if the user does not confirm the changes`, (done) => {
  const childProcess = mutagenAsync(`config import ${REJECT_CHANGES_CONFIGURATION_PATH}`, ({ code, output }) => {
    expect(output).toMatchSnapshot("changes not applied");
    expect(code).toBe(1);

    done();
  });

  confirmWith(childProcess, 'n');
}, 15000);

test(`[mutagen config import file] should import the objects if the user confirms the changes`, (done) => {
  const childProcess = mutagenAsync(`config import ${ACCEPT_CHANGES_CONFIGURATION_PATH}`, ({ code, output }) => {
    expect(output).toMatchSnapshot("changes applied");
    expect(code).toBe(0);

    done();
  });

  confirmWith(childProcess, 'y');
}, 15000);

const confirmWith = (childProcess, answer) => {
  var buffer = "";
  childProcess.stdout.on('data', (data) => {
    buffer += data;
    if (buffer.trim().endsWith("Please confirm [y|n]:")) {
      childProcess.stdin.write(`${answer}\n`)
    }
  });
}