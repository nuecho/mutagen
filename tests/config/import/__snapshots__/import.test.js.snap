// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[mutagen config import --help] should print import usage: import usage 1`] = `
Object {
  "stderr": "",
  "stdout": "Usage: import [-?] [--auto-confirm] [--trust-insecure-certificate] inputFile
[INCUBATION] Import configuration objects.
      inputFile               Input configuration file.
      --auto-confirm          Skip interactive approval before applying.
      --trust-insecure-certificate
                              Don't validate the server's TLS certificate.
  -?, -h, --help              Show this help message.
",
}
`;

exports[`[mutagen config import file] should create multiple objects of different types: every object has been created 1`] = `
Object {
  "stderr": "",
  "stdout": "+ physicalswitch [MultiObjectImportPhysicalSwitch1]
+ physicalswitch [MultiObjectImportPhysicalSwitch2]
+ switch [Environment/MultiObjectImportSwitch1]
+ switch [Environment/MultiObjectImportSwitch2]
+ switch [Environment/MultiObjectImportSwitch3]
+ dn [number: '123', switch: 'MultiObjectImportSwitch1', type: 'Extension', name: 'MultiObjectImportDN', tenant: 'Environment']
Completed. 6 object(s) imported.
",
}
`;

exports[`[mutagen config import file] should import objects even if there are dependency cycles between objects: all objects have been created 1`] = `
Object {
  "stderr": "",
  "stdout": "+ physicalswitch [CycleConfigPhysicalSwitch]
+ tenant [CycleConfigTenant]
+ script [CycleConfigTenant/CycleConfigScript]
~ tenant [CycleConfigTenant]
Completed. 3 object(s) imported.
",
}
`;

exports[`[mutagen config import file] should import the objects if the user confirms the changes: changes applied 1`] = `
Object {
  "stderr": "",
  "stdout": "The following changes are going to be applied:
+ physicalswitch [AcceptChangesPhysicalSwitch]
  {
    \\"name\\" : \\"AcceptChangesPhysicalSwitch\\",
    \\"state\\" : \\"enabled\\",
    \\"type\\" : \\"fujitsu\\"
  }
+ switch [Environment/AcceptChangesSwitch]
  {
    \\"dnRange\\" : \\"74233-74244, 74290\\",
    \\"linkType\\" : \\"madgelink\\",
    \\"name\\" : \\"AcceptChangesSwitch\\",
    \\"physicalSwitch\\" : \\"AcceptChangesPhysicalSwitch\\",
    \\"state\\" : \\"enabled\\",
    \\"switchAccessCodes\\" : [ ],
    \\"tenant\\" : \\"Environment\\"
  }
+ dn [number: '123', switch: 'AcceptChangesSwitch', type: 'Extension', name: 'AcceptChangesDN', tenant: 'Environment']
  {
    \\"name\\" : \\"AcceptChangesDN\\",
    \\"number\\" : \\"123\\",
    \\"routeType\\" : \\"default\\",
    \\"state\\" : \\"enabled\\",
    \\"switch\\" : \\"AcceptChangesSwitch\\",
    \\"switchSpecificType\\" : 1,
    \\"tenant\\" : \\"Environment\\",
    \\"trunks\\" : 1,
    \\"type\\" : \\"Extension\\",
    \\"useOverride\\" : true
  }

Please confirm [y|n]: + physicalswitch [AcceptChangesPhysicalSwitch]
+ switch [Environment/AcceptChangesSwitch]
+ dn [number: '123', switch: 'AcceptChangesSwitch', type: 'Extension', name: 'AcceptChangesDN', tenant: 'Environment']
Completed. 3 object(s) imported.
",
}
`;

exports[`[mutagen config import file] should not import any objects if some dependencies cannot be resolved and some mandatory properties for creation are missing: found missing dependencies and properties 1`] = `
Object {
  "stderr": "Validation failed.
Missing properties:
  - enumerator [Environment/emptyCfgObjectTest]:
    - displayName
    - type
Missing dependencies:
  - tenant [MissingDependencyTenant1]:
    - script [MissingDependencyTenant1/NonExistingScript1]
  - tenant [MissingDependencyTenant2]:
    - script [MissingDependencyTenant2/NonExistingScript2]
",
  "stdout": "",
}
`;

exports[`[mutagen config import file] should not import any objects if some dependencies cannot be resolved: found missing dependency 1`] = `
Object {
  "stderr": "Validation failed.
Missing dependencies:
  - tenant [MissingDependencyTenant1]:
    - script [MissingDependencyTenant1/NonExistingScript1]
  - tenant [MissingDependencyTenant2]:
    - script [MissingDependencyTenant2/NonExistingScript2]
",
  "stdout": "",
}
`;

exports[`[mutagen config import file] should not import the object if it has an invalid tenant: no object has been imported 1`] = `
Object {
  "stderr": "Validation failed.
Missing dependencies:
  - switch [Invalid tenant/switchWithInvalidTenant]:
    - tenant [Invalid tenant]
",
  "stdout": "",
}
`;

exports[`[mutagen config import file] should not import the objects if the user does not confirm the changes: changes not applied 1`] = `
Object {
  "stderr": "",
  "stdout": "The following changes are going to be applied:
+ physicalswitch [RejectChangesPhysicalSwitch]
  {
    \\"name\\" : \\"RejectChangesPhysicalSwitch\\",
    \\"state\\" : \\"enabled\\",
    \\"type\\" : \\"fujitsu\\"
  }
+ switch [Environment/RejectChangesSwitch]
  {
    \\"dnRange\\" : \\"74233-74244, 74290\\",
    \\"linkType\\" : \\"madgelink\\",
    \\"name\\" : \\"RejectChangesSwitch\\",
    \\"physicalSwitch\\" : \\"RejectChangesPhysicalSwitch\\",
    \\"state\\" : \\"enabled\\",
    \\"switchAccessCodes\\" : [ ],
    \\"tenant\\" : \\"Environment\\"
  }
+ dn [number: '123', switch: 'RejectChangesSwitch', type: 'Extension', name: 'RejectChangesDN', tenant: 'Environment']
  {
    \\"name\\" : \\"RejectChangesDN\\",
    \\"number\\" : \\"123\\",
    \\"routeType\\" : \\"default\\",
    \\"state\\" : \\"enabled\\",
    \\"switch\\" : \\"RejectChangesSwitch\\",
    \\"switchSpecificType\\" : 1,
    \\"tenant\\" : \\"Environment\\",
    \\"trunks\\" : 1,
    \\"type\\" : \\"Extension\\",
    \\"useOverride\\" : true
  }

Please confirm [y|n]: Import cancelled.
",
}
`;

exports[`[mutagen config import file] should only import objects with a unique key: only one object with the key has been imported 1`] = `
Object {
  "stderr": "",
  "stdout": "+ switch [Environment/sameSwitch]
Completed. 1 object(s) imported.
",
}
`;

exports[`[mutagen config import file] should update multiple objects of different types: every object has been updated 1`] = `
Object {
  "stderr": "",
  "stdout": "~ dn [number: '123', switch: 'MultiObjectImportSwitch1', type: 'Extension', name: 'MultiObjectImportDN', tenant: 'Environment']
~ physicalswitch [MultiObjectImportPhysicalSwitch1]
~ physicalswitch [MultiObjectImportPhysicalSwitch2]
~ switch [Environment/MultiObjectImportSwitch1]
~ switch [Environment/MultiObjectImportSwitch2]
~ switch [Environment/MultiObjectImportSwitch3]
Completed. 6 object(s) imported.
",
}
`;

exports[`[mutagen config import] should fail when file to import does not exist: no such file or directory 1`] = `
Object {
  "stderr": "/tmp/foo (No such file or directory)
",
  "stdout": "",
}
`;
